### PR TITLE
Modules

### PR DESCRIPTION
**Pull request by kubaplas**

### CODE REVIEW

File: core/src/variable.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+use crate::arithmetic::{OverflowingAdd, OverflowingSub};
```
Comment (zie1ony) [inline core/src/variable.rs]: I'd add `arithmetic`'s to prelude.

Comment (kubaplas) [reply core/src/variable.rs]: Done

File: core/src/call_result.rs [hunk 2]
```diff
@@ -5,15 +5,15 @@ use casper_event_standard::EventInstance;
 
 #[derive(Debug, Clone)]
 pub struct CallResult {
-    pub contract_address: Address,
-    pub caller: Address,
-    pub gas_used: u64,
-    pub result: Result<Bytes, OdraError>,
-    pub events: BTreeMap<Address, Vec<Bytes>>
+    pub(crate) contract_address: Address,
```
Comment (kpob) [inline core/src/call_result.rs:8]: Why are these fields pub?

File: modules/src/security/errors.rs [hunk 3]
```diff
@@ -1,8 +1,12 @@
...
+use odra::OdraError;
 
-execution_error! {
-    pub enum Error {
-        PausedRequired => 21_000,
-        UnpausedRequired => 21_001,
+pub enum Error {
+    PausedRequired = 21_000,
+    UnpausedRequired = 21_001
+}
```
Comment (kpob) [inline modules/src/security/errors.rs]:
```suggestion
#[derive(OdraError)]
pub enum Error {
    PausedRequired = 21_000,
    UnpausedRequired = 21_001
}
```

File: modules/src/access/errors.rs [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
-        CallerNotTheNewOwner => 20_002,
-        MissingRole => 20_003,
-        RoleRenounceForAnotherAddress => 20_004,
+pub enum Error {
+    OwnerNotSet = 20_000,
+    CallerNotTheOwner = 20_001,
+    CallerNotTheNewOwner = 20_002,
+    MissingRole = 20_003,
+    RoleRenounceForAnotherAddress = 20_004
+}
```
Comment (kpob) [inline modules/src/access/errors.rs]:
```suggestion
#[derive(OdraError)]
pub enum Error {
    OwnerNotSet = 20_000,
    CallerNotTheOwner = 20_001,
    CallerNotTheNewOwner = 20_002,
    MissingRole = 20_003,
    RoleRenounceForAnotherAddress = 20_004
}
```

File: modules/src/erc1155/erc1155_base.rs [hunk 5]
```diff
@@ -1,33 +1,26 @@
...
 }
 
 impl Erc1155 for Erc1155Base {
     fn balance_of(&self, owner: &Address, id: &U256) -> U256 {
-        self.balances.get_instance(owner).get_or_default(id)
+        self.balances.get_or_default(&(*owner, *id))
     }
 
-    fn balance_of_batch(&self, owners: &[Address], ids: &[U256]) -> Vec<U256> {
+    fn balance_of_batch(&self, owners: Vec<Address>, ids: Vec<U256>) -> Vec<U256> {
```
Comment (kpob) [inline modules/src/erc1155/erc1155_base.rs:21]: Why did you change a slice to a vec? Does it work with slices?

File: modules/src/erc721/extensions/erc721_metadata.rs [hunk 6]
```diff
@@ -45,13 +53,17 @@ impl Erc721MetadataExtension {
...
 
 pub mod errors {
-    use odra::execution_error;
+    use odra::OdraError;
+
+    pub enum Error {
+        NameNotSet = 31_000,
+        SymbolNotSet = 31_001,
+        BaseUriNotSet = 31_002
+    }
```
Comment (kpob) [inline modules/src/erc721/extensions/erc721_metadata.rs]:
```suggestion
    #[derive(OdraError)]
    pub enum Error {
        NameNotSet = 31_000,
        SymbolNotSet = 31_001,
        BaseUriNotSet = 31_002
    }
```

File: modules/src/erc721/mod.rs [hunk 7]
```diff
@@ -81,21 +82,25 @@ pub mod events {
 
 /// Erc721-related Odra errors.
 pub mod errors {
-    use odra::execution_error;
+    use odra::OdraError;
+
+    /// Possible errors in the context of Erc721 token.
+    pub enum Error {
```
Comment (kpob) [inline modules/src/erc721/mod.rs:89]:
```suggestion
    #[derive(OdraError)]
    pub enum Error {
```

File: modules/src/erc721_receiver.rs [hunk 8]
```diff
@@ -1,45 +1,45 @@
...
-use odra::types::U256;
-use odra::types::{event::OdraEvent, Address};
+use crate::erc721_token::Erc721TokenContractRef;
+use odra::prelude::*;
+use odra::{Address, Bytes, Module, U256};
 
 /// The ERC721 receiver implementation.
 #[odra::module]
-pub struct Erc721Receiver;
+pub struct Erc721Receiver {}
```
Comment (kpob) [inline modules/src/erc721_receiver.rs:11]: Add a TODO, #295 resolves this issue. If you merge first, I'll fix it then

**Review by kpob — APPROVED** [review]

File: modules/src/erc1155/erc1155_base.rs (hunk 5 above)
Comment (kubaplas) [reply modules/src/erc1155/erc1155_base.rs:21]: It did not in some place, so I changed it everywhere

File: core/src/call_result.rs (hunk 2 above)
Comment (kubaplas) [reply core/src/call_result.rs:8]: HostEnv needs to construct CallResult.
